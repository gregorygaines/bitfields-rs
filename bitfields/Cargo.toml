[package]
name = "bitfields"
version = "0.1.0"
authors = ["Gregory Gaines <gainesagregory@gmail.com>"]
description = "Macro for generating flexible bitfields. Useful for low-level code (embedded or emulators)."
keywords = ["bitfield", "proc_macro", "procmacro", "struct", "macro"]
edition = "2021"

[dependencies]
bitfields-proc-macro = { path = "../bitfields_proc_macro" }

[dev-dependencies]
trybuild = "1.0.101"
