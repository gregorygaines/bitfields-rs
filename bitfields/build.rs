use std::path::Path;

/// Out dir infix for Rust being built by a publish command.
const PUBLISH_OUT_DIR_INFIX: &str = "target/package/";

/// I have doc tests in the README file, but the README file is outside the
/// bitfields crate and cargo refuses to include it in the crate. So I copy the
/// README file to the target directory and include it from there.
fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Get the target directory
    let out_dir = std::env::var("OUT_DIR")?;

    // Source path of the README file
    let src_path = if out_dir.contains(PUBLISH_OUT_DIR_INFIX) {
        println!("cargo:rerun-if-changed=../../../README.md");
        Path::new("../../../README.md")
    } else {
        println!("cargo:rerun-if-changed=../README.md");
        Path::new("../README.md")
    };

    // Destination path in the target directory
    let dst_path = Path::new(&out_dir).join("README.md");

    // Copy the README to the target directory
    std::fs::copy(src_path, dst_path)?;

    Ok(())
}
