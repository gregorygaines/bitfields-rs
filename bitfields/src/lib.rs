#![doc = include_str!(concat!(env!("OUT_DIR"), "/README.md"))]

pub use bitfields_impl::bitfield;
