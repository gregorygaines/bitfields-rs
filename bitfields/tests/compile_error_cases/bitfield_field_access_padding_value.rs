use bitfields::bitfield;

#[bitfield(u64)]
pub struct Bitfield {
    #[bits(4, default = 0x1)]
    a: u8,
    #[bits(60, default = 0xFFFF_FFFF_FFFF)]
    _padding: u64
}

fn main() {
    let bitfield = Bitfield::new();
    assert_eq!(bitfield.a(), 0x1);
    assert_eq!(bitfield.into_bits(), 0xF_FFFF_FFFF_FFF1);
    assert_eq!(bitfield._padding(), 0x1);
}
