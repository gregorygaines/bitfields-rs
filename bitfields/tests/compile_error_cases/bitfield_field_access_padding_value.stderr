error[E0599]: no method named `_padding` found for struct `Bitfield` in the current scope
  --> tests/compile_error_cases/bitfield_field_access_padding_value.rs:15:25
   |
 3 | #[bitfield(u64)]
   | ---------------- method `_padding` not found for this struct
...
15 |     assert_eq!(bitfield._padding(), 0x1);
   |                         ^^^^^^^^ method not found in `Bitfield`
